docs/proposals/20201020-management-cluster-operator.md

```diff
@@ -0,0 +1,1063 @@
+`spec.Paused` field to true for each provider; the operator will block any
+contract upgrade until all the providers are paused.
+
+After all the providers are in paused state, you can proceed with the upgrade
+as described in the previous paragraph (change the `spec.Version` field).
+
+When a provider is paused the number of replicas will be scaled to 0; the
+operator will add a new
+`management.cluster.x-k8s.io/original-controller-replicas` annotation to store
+the original replica count.
+
+Once all the providers are upgraded to a version that abides to the new
+contract, it is possible for the operator to unpause providers; the operator
+does not allow to unpause providers if there are still providers abiding to
+the old contract.
+
+Please note that we are planning to embed this sequence (pause - upgrade -
+unpause) as a part of `clusterctl upgrade apply` command when there is a
+contract change.
+
+##### Changing a provider
+
+On top of changing a provider version (upgrades), the operator supports also
+changing other provider fields, most notably controller flags and variables.
+This can be acheived by either `kubectl edit` or `kubectl apply` to the
+provider object.
+
+The operation internally works like upgrades: The current instance of the
+provider is deleted, while preserving CRDs, namespaced and user objects A new
+instance of the provider is installed with the new set of flags/variables.
+
+Please note that clusterctl currently does not support this operation.
+
+See Example 1 in [Example API Usage](#example-api-usage)
+
+##### Deleting a provider
+
+In order to delete a provider you have to delete the corresponding provider
+object.
+
+Deletion of the provider will be blocked if any workload cluster using the
+provider still exists.
+
+Additionally, deletion of a core provider should be blocked if there are still
+other providers in the management cluster.
+
+#### Upgrade from v1alpha3 management cluster to v1alpha4 cluster
+
+Cluster API will provide instructions on how to upgrade from a v1alpha3
+management cluster, created by clusterctl to the new v1alpha4 management
+cluster. These operations could require manual actions.
+
+Some of the actions are described below:
+- Run webhooks as part of the main manager. See [issue 3822].
+
+More details will be added as we better understand what a v1alpha4 cluster
+will look like.
+
+#### Operator Lifecycle Management
+
+##### Operator Installation
+
+- `clusterctl init` will install the operator and its corresponding CRDs as a
+  pre-requisite if the operator doesn’t already exist. Please note that this
+  command will consider image overrides defined in the local clusterctl config
+  file.
+- If the admin does not want to use clusterctl to install the operator, it is
+  possible to kubectl apply the operator yaml that will be published in the
+  cluster-api release artifacts.
+- There can be only one operator installed per management cluster, across all
```

Comment:
how is this enforced?

Reply:
Good point. `clusterctl` will enforce this if the users use `clusterctl init`. 
If the user installs the operator manually (kubectl), then we rely on documentation? 🤷🏽 
🤔  I guess the operator itself could do some sort of check. That is, if installing the second instance of the operator, it can determine if there is an operator already installed. Not sure about the implementation. But I'm open to suggestions.

Reply:
I'm +1 to remove this sentence given that it could be confusing.

We are going to support many running replicas of the operator, but only one will be active via leader election (e.g for HA purposes).

What is open for discussion is if this operator should support the --namespace flag (for reconciling objects in a single namespace), but this is strictly linked to the fact of using global or namespaced CRDs.

Reply:
Ok. I'll remove this line since it is a little confusing.